src/plugin/session-tracker.ts

```diff
@@ -0,0 +1,182 @@
+    public readonly onSessionEvent = this._onSessionEvent.event;
+
+    activate(context: vscode.ExtensionContext): void {
+        context.subscriptions.push(
+            vscode.debug.registerDebugAdapterTrackerFactory('*', this),
+            vscode.debug.onDidChangeActiveDebugSession(session => this.activeSessionChanged(session))
+        );
+    }
+
+    createDebugAdapterTracker(session: vscode.DebugSession): vscode.ProviderResult<vscode.DebugAdapterTracker> {
+        return ({
+            onWillStartSession: () => this.sessionWillStart(session),
+            onWillStopSession: () => this.sessionWillStop(session),
+            onDidSendMessage: message => this.adapterMessageReceived(session, message),
+            onWillReceiveMessage: message => this.willSendClientMessage(session, message)
+        });
+    }
+
+    protected sessionInfo(session: vscode.DebugSession): SessionInfo {
+        let info = this._sessionInfo.get(session.id);
+        if (!info) {
+            info = { raw: session };
+            this._sessionInfo.set(session.id, info);
+        }
+        return info;
+    }
+
+    protected async activeSessionChanged(session?: vscode.DebugSession): Promise<void> {
+        for (const [sessionId, info] of this._sessionInfo) {
+            info.active = sessionId === session?.id;
+        }
+        this._onSessionEvent.fire({ event: 'active', session: session ? this.sessionInfo(session) : undefined });
+    }
+
+    protected fireSessionEvent<K extends keyof Omit<SessionEvents, 'active'>>(session: vscode.DebugSession, event: K, data: SessionEvents[K]['data']): void {
+        this._onSessionEvent.fire({ event, session: this.sessionInfo(session), data });
+    }
+
+    protected async sessionWillStart(session: vscode.DebugSession): Promise<void> {
+        this._sessionInfo.set(session.id, { raw: session });
+    }
+
+    protected sessionWillStop(session: vscode.DebugSession): void {
+        this._sessionInfo.delete(session.id);
+    }
+
+    protected willSendClientMessage(session: vscode.DebugSession, message: unknown): void {
+        if (isDebugRequest('initialize', message)) {
+            this.sessionInfo(session).clientCapabilities = message.arguments;
+        }
+    }
+
+    protected adapterMessageReceived(session: vscode.DebugSession, message: unknown): void {
+        if (isDebugResponse('initialize', message)) {
+            this.sessionInfo(session).debugCapabilities = message.body;
+        } else if (isDebugEvent('stopped', message)) {
+            this.sessionInfo(session).stopped = true;
+            this.fireSessionEvent(session, 'stopped', undefined);
+        } else if (isDebugEvent('continued', message)) {
+            this.sessionInfo(session).stopped = false;
+            this.fireSessionEvent(session, 'continued', undefined);
+        } else if (isDebugEvent('memory', message)) {
+            this.fireSessionEvent(session, 'memory-written', message.body);
```

Comment:
This illustrates some of the difficulty with the division of labor. This isn't really a matter of status, and the rest of the bookkeeping for memory retrieval and writing is handled in the memory provider. I think it would be preferable to keep this in the memory provider, but I don't feel very strongly about it.

Reply:
I see your point. However, I'm not sure we should see it strictly as 'state management' vs 'memory management' but rather 'enriched session management' by the tracker (e.g., more events could be added in the future). Similar to the npm package you shared above, I'd almost argue that the session tracker is just one layer in-between and the reason the memory-provider also needs to register as a debug tracker is because we want to allow external contributions.